source/agora/common/Hash.d

```diff
@@ -289,3 +289,182 @@ nothrow @nogc @safe unittest
+    As a result, we can compress the data as much as we want,
+    to reduce memory usage. However, this comes at the cost of more computation.
+
+    For example, if we have cycles of 10 pre-images, we can first compute the
+    seed (H0), then reveal our initial commitment (H9). This will cost us
+    9 hash operations. Then on the next reveal, in order to get H8, we need to
+    perform 8 hash operations, then 7 for H7, etc...
+
+    If we want to improve this, we can store intermediate results:
+    when generating H9, we do perform 9 hash operations, but store H5.
+    Hence, when generating H8 and H7, we only need to perform
+    3 and 2 hash operations, respectively, and will never need to perform more
+    than 4 operations (save the initial commitment).
+
+    This structure is a mean to generalize this approach, with arbitrary count
+    and arbitrary sample size (interval, or distance, between saved pre-images).
+
+*******************************************************************************/
+
+public struct PreImageCache
+{
+    @safe nothrow:
+
+    /// Store the actual preimages
+    private Hash[] data;
+
+    /// Interval between two preimages in `data`
+    private const ulong interval;
+
+    /// Construct an instance using already existing data
+    public this (inout(Hash)[] data_, ulong sample_size) inout @nogc pure
+    {
+        this.data = data_;
+        this.interval = sample_size;
+    }
+
+    /***************************************************************************
+
+        Construct an instance and allocate memory
+
+        This takes a `count` of pre-image and a sample size, or distance.
+        For example, if one wish to store 1000 pre-images in this cache,
+        and perform no more than 5 hash operations each time,
+        the `sample_size` should be `5` and the `count` should be `200`
+        (1000 / 5).
+
+        Params:
+          count = Number of samples to store (number of entries in the array)
+          sample_size = Distance between two pre-images
+
+    ***************************************************************************/
+
+    public this (ulong count, ulong sample_size) pure
+    {
+        assert(count > 1, "A count of less than 2 does not make sense");
+        assert(sample_size > 1, "A distance of less than 2 does not make sense");
+
+        this.interval = sample_size;
+        this.data = new Hash[](count);
+    }
+
+    /***************************************************************************
+
+        Populate this cache from the `seed`
+
+        Params:
+          seed = Initial value to derive preimage from. This will always be the
+                 first value of the array.
+          length = The number of entries to populate.
+                   This can be used when large sample size are used,
+                   and one wish to stop initialization past a certain threshold.
+                   The special value `0` means "everything", and is the default.
```

Comment:
I wonder if overloads would be better fit for this? For example:

```D
/// Returns: The result of hashing seed `sample_size * (count + 1) - 1`
public Hash reset (Hash seed)
{
    return this.reset(seed, this.data.length);
}

/// Returns: The result of hashing seed `sample_size * (length + 1) - 1`
public Hash reset (Hash seed, size_t length)
{
    ...
}
```

It's a little easier to reason about having two different functions rather than 1 parameter that has a different meaning based on its value.

Reply:
Good point. I wasn't very happy with this API, and your suggestion looks very good. One point though: The `reset(Hash seed)` should be called by `reset(Hash, size_t)`, this way the former can be made `@nogc`.